server/src/main/java/org/opensearch/wlm/cancellation/DefaultTaskCancellation.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.wlm.cancellation;
+
+import org.opensearch.cluster.metadata.QueryGroup;
+import org.opensearch.monitor.jvm.JvmStats;
+import org.opensearch.monitor.process.ProcessProbe;
+import org.opensearch.search.ResourceType;
+import org.opensearch.tasks.TaskCancellation;
+import org.opensearch.wlm.QueryGroupLevelResourceUsageView;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BooleanSupplier;
+import java.util.stream.Collectors;
+
+import static org.opensearch.wlm.tracker.QueryGroupResourceUsageTrackerService.TRACKED_RESOURCES;
+
+/**
+ * Manages the cancellation of tasks enforced by QueryGroup thresholds on resource usage criteria.
+ * This class utilizes a strategy pattern through {@link DefaultTaskSelectionStrategy} to identify tasks that exceed
+ * predefined resource usage limits and are therefore eligible for cancellation.
+ *
+ * <p>The cancellation process is initiated by evaluating the resource usage of each QueryGroup against its
+ * resource limits. Tasks that contribute to exceeding these limits are selected for cancellation based on the
+ * implemented task selection strategy.</p>
+ *
+ * <p>Instances of this class are configured with a map linking QueryGroup IDs to their corresponding resource usage
+ * views, a set of active QueryGroups, and a task selection strategy. These components collectively facilitate the
+ * identification and cancellation of tasks that threaten to breach QueryGroup resource limits.</p>
+ *
+ * @see DefaultTaskSelectionStrategy
+ * @see QueryGroup
+ * @see ResourceType
+ */
+public class DefaultTaskCancellation {
+    private static final long HEAP_SIZE_BYTES = JvmStats.jvmStats().getMem().getHeapMax().getBytes();
+
+    protected final DefaultTaskSelectionStrategy defaultTaskSelectionStrategy;
+    // a map of QueryGroupId to its corresponding QueryGroupLevelResourceUsageView object
+    protected final Map<String, QueryGroupLevelResourceUsageView> queryGroupLevelResourceUsageViews;
+    protected final Set<QueryGroup> activeQueryGroups;
+    protected BooleanSupplier isNodeInDuress;
+
+    public DefaultTaskCancellation(
+        DefaultTaskSelectionStrategy defaultTaskSelectionStrategy,
+        Map<String, QueryGroupLevelResourceUsageView> queryGroupLevelResourceUsageViews,
+        Set<QueryGroup> activeQueryGroups,
+        BooleanSupplier isNodeInDuress
+    ) {
+        this.defaultTaskSelectionStrategy = defaultTaskSelectionStrategy;
+        this.queryGroupLevelResourceUsageViews = queryGroupLevelResourceUsageViews;
+        this.activeQueryGroups = activeQueryGroups;
+        this.isNodeInDuress = isNodeInDuress;
+    }
+
+    /**
+     * Cancel tasks based on the implemented strategy.
+     */
+    public final void cancelTasks() {
```

Comment:
Since we haven't handled the delete queryGroups yet, I thought I'd address that there. 
But I think addressing that here makes sense, look out for it in the next commits, good call out !